### PR TITLE
Use a distinct error code for "if may be missing an else clause"

### LINKED ISSUE #36596: Error E0308 isn't relevant to the original error message
Here's my code (2.rs):

```rust
fn main() {
    let x = 5;
    let a = if x == 5 { 1 };
    println!("{}", a);
}
```

On running `rustc 2.rs`, I get

```
2.rs:3:13: 3:28 error: if may be missing an else clause [E0308]
2.rs:3     let a = if x == 5 { 1 };
                   ^~~~~~~~~~~~~~~
2.rs:3:13: 3:28 help: run `rustc --explain E0308` to see a detailed explanation
2.rs:3:13: 3:28 note: expected type `()`
2.rs:3:13: 3:28 note:    found type `_`
error: aborting due to previous error
```

And `rustc --explain E0308` gives

```
This error occurs when the compiler was unable to infer the concrete type of a
variable. It can occur for several cases, the most common of which is a
mismatch in the expected type that the compiler inferred for a variable's
initializing expression, and the actual type explicitly assigned to the
variable.

For example:

let x: i32 = "I am not a number!";

...
````

Am I missing something? Shouldn't the explanation talk about `error: if may be missing an else clause`?

### PR DESCRIPTION
**Pull request by jfirebaugh**
Introduce the possibility of assigning distinct error codes to the various origin types of E0308. Start by assigning E0317 for the "IfExpressionWithNoElse" case, and write a long diagnostic specific to this case.

Fixes #36596

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1431,6 +1431,21 @@ fn make_child<'elve>(x: &mut &'elve isize, y: &mut &'elve isize) {
...
 "##,
 
+E0317: r##"
+An `if` expression without an `else` block is required to have the type `()`.
+This error occurs when the `if` block has a type other than `()`. For example:
+
+```compile_fail,E0317
+fn main() {
+    let x = 5;
+    let a = if x == 5 { 1 };
```
Comment (nagisa) [inline src/librustc/diagnostics.rs]: This example can be made more relatable if written as such:

```
if condition {
    some_value
}
```

Missing semicolon after last expression in an `if` is likely the most common reason for this error by a large degree.

Comment (jfirebaugh) [inline src/librustc/diagnostics.rs]: That example produces the generic "mismatched types" variant of E0308 though, not "if may be missing an else clause".

Comment (nagisa) [inline src/librustc/diagnostics.rs]: Does it? That’s weird. I would expect `else`-less `if` as a last expression in a function body/block to produce the “if may be missing an else clause” error instead.

Comment (jfirebaugh) [inline src/librustc/diagnostics.rs]: It seems to be dependent on the expected type in the expression's context:

```
fn foo() -> i32 {
    if true {
        5
    }
}
```

produces:

```
error[E0308]: if may be missing an else clause
 --> <anon>:2:5
  |
2 |     if true {
  |     ^ expected (), found i32
  |
```

However, change `i32` to `()` and it's:

```
error[E0308]: mismatched types
 --> <anon>:3:9
  |
3 |         5
  |         ^ expected (), found integral variable
  |
```

Comment (nikomatsakis) [inline src/librustc/diagnostics.rs]: This makes sense to me -- even if you added an else, given that the expected type of this expression is `()`, the body would have to be changed. That is, the body and the fn signature are in conflict. We choose to blame the body. You could make a case the other way, but blaming the *if* seems wrong, since there is nothing you can change about the `if` itself (without changing something else) that would make this code compile.

Not that the compiler thinks this hard when selecting what to blame; it's semi-random. Just saying that in this case the *span* makes sense. It'd be nice if we could give a better error message though, explaining where the `()` came from.

I think that Elm does some nice tracking here. We should copy them. =)

**Comment by jfirebaugh** [comment]
Any additional feedback or changes needed here? I have some work ready that builds on this but I'd like to get this PR merged first.

File: src/librustc/infer/error_reporting.rs [hunk 2]
```diff
@@ -577,11 +577,17 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
     {
-        // FIXME: do we want to use a different error code for each origin?
-        let mut diag = struct_span_err!(
-            self.tcx.sess, trace.origin.span(), E0308,
-            "{}", trace.origin.as_failure_str()
-        );
+        let span = trace.origin.span();
+        let failure_str = trace.origin.as_failure_str();
+        let mut diag = match trace.origin {
+            // FIXME: use distinct codes for each case
```
Comment (nikomatsakis) [inline src/librustc/infer/error_reporting.rs]: I'd say let's just remove this FIXME, but if not, reword it to be more tentative, like "FIXME: Do we want to have distinct codes for other cases?"

**Comment by nikomatsakis** [comment]
@jfirebaugh looks good, sorry for the delay.

**Comment by nikomatsakis** [comment]
@jfirebaugh er, sorry, I meant "looks good apart from that nit." Think you can tweak that? If so, let me know and I'll r+

**Comment by jfirebaugh** [comment]
@nikomatsakis Agreed; removed the FIXME.
